Mesh_3/include/CGAL/Labeled_mesh_domain_3.h

```diff
@@ -37,23 +37,217 @@
+      return functor;
+    }
+  };
+
+  // specialization for `Null_functor`: create the default functor
+  template <>
+  struct Create_labeled_image_values_to_subdomain_indices<Null_functor> {
+    typedef Identity type;
+    type operator()(Null_functor) const {
+      return type();
+    }
+  };
+
+} // end namespace CGAL::internal::Mesh_3
+} // end namespace CGAL::internal
+
 struct Null_subdomain_index {
   template <typename T>
   bool operator()(const T& x) const { return 0 == x; }
 };
 
+template <typename Subdomain_index>
+struct Construct_pair_from_subdomain_indices {
+  typedef std::pair<Subdomain_index, Subdomain_index> result_type;
+
+  result_type operator()(Subdomain_index a, Subdomain_index b) const {
+    return result_type(a, b);
+  }
+}; // end class template Construct_pair_from_subdomain_indices
+
+template <typename Geom_traits,
+          typename Subdomain_index,
+          typename Surface_patch_index_>
+class Labeled_mesh_domain_3_impl_details {
+protected:
+  typedef Surface_patch_index_ Surface_patch_index;
+  typedef typename Geom_traits::Point_3 Point_3;
+  typedef typename Geom_traits::Iso_cuboid_3 Iso_cuboid_3;
+  typedef typename Geom_traits::FT FT;
+  typedef boost::shared_ptr<CGAL::Random> CGAL_Random_share_ptr_t;
+  /// Returns squared error bound from \c bbox and \c error
+  FT squared_error_bound(const Iso_cuboid_3& bbox, const FT& error) const
+  {
+    typename Geom_traits::Compute_squared_distance_3 squared_distance =
+                             Geom_traits().compute_squared_distance_3_object();
+    return squared_distance((bbox.min)(), (bbox.max)())*error*error/4;
+  }
+
+  /// Returns and Iso_cuboid_3 from a Bbox_3
+  static Iso_cuboid_3 iso_cuboid(const Bbox_3& bbox)
+  {
+    const Point_3 p_min(bbox.xmin(), bbox.ymin(), bbox.zmin());
+    const Point_3 p_max(bbox.xmax(), bbox.ymax(), bbox.zmax());
+
+    return Iso_cuboid_3(p_min,p_max);
+  }
+
+  /// Returns and Iso_cuboid_3 from a Bbox_3
+  static Iso_cuboid_3 iso_cuboid(const typename Geom_traits::Sphere_3& sphere)
+  {
+    return iso_cuboid(sphere.bbox());
+  }
+
+  /// Returns and Iso_cuboid_3 from a Bbox_3
+  static Iso_cuboid_3 iso_cuboid(const typename Geom_traits::Iso_cuboid_3& c)
+  {
+    return c;
+  }
+
+  static Construct_pair_from_subdomain_indices<Subdomain_index>
+  construct_pair_functor() {
+    return Construct_pair_from_subdomain_indices<Subdomain_index>();
+  }
+
+  template <class ArgumentPack>
+  Labeled_mesh_domain_3_impl_details(ArgumentPack const& args)
+    : function_(args[parameters::function])
+    , bbox_(iso_cuboid(args[parameters::bounding_object]))
+    , cstr_s_p_index(args[parameters::construct_surface_patch_index |
+                          construct_pair_functor()])
+    , null(args[parameters::null_subdomain_index | Null_subdomain_index()])
+    , p_rng_(args[parameters::p_rng|0] == 0 ?
+             CGAL_Random_share_ptr_t(new CGAL::Random(0)) :
```

Comment:
Naive question and the doc is not very talkative: how does `Random(0)` compare to `get_default_random()`? Why is it rather initialized to `0` here?

Reply:
That is not documented, but`get_default_random()`returns a default-constructed `Random` object. And that means [it is seeded based on the system item of its creation][Random()].

Whereas `Random(0)` is a `Random` object seeded with the constant `0`.

As we have users of Mesh_3 who do not like non-determinism... we seed from `0`.

[Random()]: https://doc.cgal.org/4.12/Generator/classCGAL_1_1Random.html#aa200d349235965500be514f6d6a566e1

Reply:
I know about the seeding, I meant maybe from a concurrency point of view or something like that (it's very possible that the question does not make sense then).

Reply:
`p_rng_` was initialized by a pointer to the object returned by `get_default_random()`, then the class would no longer be thread-safe. The `get_default_random()` is thread-safe in the sense that it returns a thread-local variable, and each thread gets a reference to a different object.

But if you create a `Labeled_mesh_domain_3` is the main thread, store the reference to the default `Random` object, and then use Mesh_3 with TBB and the parallel mode, then all thread would use the same `Random` object, that is not thread-safe!

Reply:
Sorry for the typos in my message. I fixed most of them after a proof-reading of my message.

Reply:
Right, thanks for the clarification.